### PR TITLE
Handle EmailOTP error response using internal errorcodes

### PR DESCRIPTION
**Pull request by piraveena**
Fix https://github.com/wso2/product-is/issues/7569

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 1]
```diff
@@ -1266,5 +1278,15 @@ protected void triggerNotification(String userName, String tenantDomain, String
...
 
+    private String getHttpErrorResponseCode(String errorMsg){
 
-}
\ No newline at end of file
+        String errorCode;
+        if (StringUtils.contains(errorMsg, ":")) {
+            String[] code = errorMsg.split(":");
+            errorCode = code[0].replace(" ", "");
+        }else{
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: formatting issue

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 2]
```diff
@@ -1266,5 +1278,15 @@ protected void triggerNotification(String userName, String tenantDomain, String
         }
     }
 
+    private String getHttpErrorResponseCode(String errorMsg){
 
-}
\ No newline at end of file
+        String errorCode;
+        if (StringUtils.contains(errorMsg, ":")) {
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Can't we use StringUtils.replace(text, searchString, replacement)

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java [hunk 3]
```diff
@@ -41,6 +42,7 @@
 public class SMSOTPUtils {
 
     private static final Log log = LogFactory.getLog(SMSOTPUtils.class);
+    private static boolean useInternalErrorCodes = true;
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java]: shouldn't the default value be false for this?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java [hunk 4]
```diff
@@ -356,6 +358,30 @@ public static String getBackupCode(AuthenticationContext context) {
...
+     * Check whether using int ernal error codes is supported.
+     *
+     * @param context Authentication Context.
+     * @return Ture if useInternalError codes is enabled, else return false.
+     */
+    public static boolean useInternalErrorCodes(AuthenticationContext context) {
+
+        String useSMSProviderCodesConfig = getConfiguration(context, SMSOTPConstants.USE_INTERNAL_ERROR_CODES);
+        if (StringUtils.isNotEmpty(useSMSProviderCodesConfig)) {
+            useInternalErrorCodes = Boolean.parseBoolean(useSMSProviderCodesConfig);
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java]: let's add a debug log here

Comment (piraveena) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java]: added

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 5]
```diff
@@ -1165,7 +1175,11 @@ public boolean sendRESTCall(AuthenticationContext context, String smsUrl, String
             smsProviderUrl = new URL(smsUrl);
         } catch (MalformedURLException e) {
             log.error("Error while parsing SMS provider URL: " + smsUrl, e);
-            context.setProperty(SMSOTPConstants.ERROR_CODE, "The SMS URL does not conform to URL specification");
+            if (SMSOTPUtils.useInternalErrorCodes(context)) {
+                context.setProperty(SMSOTPConstants.ERROR_CODE, SMSOTPConstants.ErrorMessage.MAL_FORMED_URL.getCode());
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: MAL_FORMED should be a single word MALFORMED_URL

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java [hunk 6]
```diff
@@ -1266,5 +1280,12 @@ protected void triggerNotification(String userName, String tenantDomain, String
...
 
+    private String getHttpErrorResponseCode(String errorMsg) {
 
-}
\ No newline at end of file
+        String errorCode = errorMsg;
+        if (StringUtils.contains(errorCode, ":")) {
+            errorCode = errorCode.split(":")[0];
+        }
+        return errorCode.trim();
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: check whether errorCode is null before calling trim(). Check whether there's a trim function in StringUtils then we can directly use that

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPConstants.java [hunk 7]
```diff
@@ -19,8 +19,13 @@
 
 package org.wso2.carbon.identity.authenticator.smsotp;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ResourceBundle;
+
 public class SMSOTPConstants {
 
+    private static final String APPLICATION_MANAGEMENT_PREFIX = "SMSOTP-";
```
Comment (emswbandara) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPConstants.java]: change this to SMSOTP_PREFIX

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 6 above)
Comment (piraveena) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: Null check is already handled here https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/pull/81/files/60fd47f50a17eca7eded2ab24babaec65530dbd7#diff-862199ef8d38f6df18c5f44aeeb2e115R575

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java [hunk 8]
```diff
@@ -356,6 +358,33 @@ public static String getBackupCode(AuthenticationContext context) {
 
     }
 
+    /**
+     * Check whether using int ernal error codes is supported.
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPUtils.java]: Check and change the method description

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java (hunk 6 above)
Comment (piraveena) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java]: StringUtils is used

**Pull request merged**